Source/Search/Mixins/AdvancedSearchProviderMixin.lua

```diff
@@ -0,0 +1,225 @@
+  elseif eventName == "AUCTION_HOUSE_BROWSE_FAILURE" then
+    AuctionHouseFrame.BrowseResultsFrame.ItemList:SetCustomError(
+      RED_FONT_COLOR:WrapTextInColorCode(ERR_AUCTION_DATABASE_ERROR)
+    )
+  end
+end
+
+function AuctionatorAdvancedSearchProviderMixin:ProcessSearchResults(addedResults)
+  Auctionator.Debug.Message("AuctionatorAdvancedSearchProviderMixin:ProcessSearchResults()")
+
+  local results = {}
+
+  for index = 1, #addedResults do
+    -- Run filter checks on every item key. Some might not be added to the
+    -- results yet, but when the relevant information arrives in an event
+    if self:FilterByItemLevel(addedResults[index].itemKey) then
+      if self:FilterByExact(addedResults[index].itemKey) then
+        table.insert(results, addedResults[index].itemKey)
+      end
+    end
+  end
+
+  self:AddResults(results)
+end
+
+function AuctionatorAdvancedSearchProviderMixin:ProcessItemKeyInfo(itemID)
+  --Event for missing info received about itemID.
+  for index, itemKey in ipairs(self.itemKeyInfoQueue) do
+    local itemKeyInfo = C_AuctionHouse.GetItemKeyInfo(itemKey)
+    if itemKeyInfo then
+      --Remove key from list of those with missing info
+      table.remove(self.itemKeyInfoQueue, index)
+
+    --Only exact search uses this info, and the event won't have been queued
+    --otherwise.
+      if self:ExactMatchCheck(itemKeyInfo) then
+        self:AddResults({itemKey})
+      else
+        --Post empty results, so the mixin supplying it runs
+        --self:HasCompleteTermResults() and can see if the search is complete
+        self:AddResults({})
+      end
+
+      --Only one new result per event
+      break
+    end
+  end
+end
+
+function AuctionatorAdvancedSearchProviderMixin:FilterByItemLevel(itemKey)
+  -- Check for 0 is to avoid filtering issues with glitchy AH APIs.
+  if itemKey.itemLevel ~= nil and itemKey.itemLevel ~= 0 then
+    --Minimum item level check
+    if (self.currentFilter.minItemLevel ~= nil and self.currentFilter.minItemLevel>itemKey.itemLevel) then
+      return false
+    end
+    --Maximum item level check
+    if (self.currentFilter.maxItemLevel ~= nil and self.currentFilter.maxItemLevel<itemKey.itemLevel) then
+      return false
+    end
+  end
+  return true
+end
+
+function AuctionatorAdvancedSearchProviderMixin:FilterByExact(itemKey)
+  if self.currentFilter.exactSearch ~= nil then
+    local itemKeyInfo = C_AuctionHouse.GetItemKeyInfo(itemKey)
+
+    if itemKeyInfo == nil then
+      Auctionator.Debug.Message("AuctionatorAdvancedSearchProviderMixin:FilterByExact Missing itemKeyInfo")
+
+      --Put key in the queue for completing filtering later in an
+      --ITEM_KEY_ITEM_INFO_RECEIVED event
+      table.insert(self.itemKeyInfoQueue, itemKey)
+
+      return false
+    else
+      Auctionator.Debug.Message("AuctionatorAdvancedSearchProviderMixin:FilterByExact Got itemKeyInfo", itemKeyInfo.itemName)
+
+      return self:ExactMatchCheck(itemKeyInfo)
+    end
+  end
+  return true
```

Comment:
Add whitespace above this line please (to make it clear where the if block ends and where the default return begins).